Gzip payload before upload

vigneshashokan opened the pull request:
Compress (gzip) payload before upload.

Commit:
gzip payload before upload

Commit:
Merge branch 'master' into gzip-payload

Commit:
fix merge issues

mestroma commented on hook.go:
```diff
@@ -135,11 +137,23 @@ func (h *SumoLogicHook) queue(msg SumoLogicMesssage) (err error) {
+	var buf bytes.Buffer
+	g := gzip.NewWriter(&buf)
+	if _, err := g.Write(payload); err != nil {
+		return err
```
I think it will be good to log error so user is aware what went wrong with uploading log
`h.errorf("compressing  payload - %s", err)`

mestroma commented on hook.go:
```diff
@@ -135,11 +137,23 @@ func (h *SumoLogicHook) queue(msg SumoLogicMesssage) (err error) {
+	}
+
+	if err := g.Close(); err != nil {
+		return err
```
same like above log error

mestroma reviewed (changes requested):
Thanks for the PR.
Can you please fix the tests?
Can you please set appropriate header for gzip `req.Header.Set("Content-Encoding", "gzip")`
I think it will be good if compression can be turned off optionally in configuration. In case someone wants real time logs and they limit batch size to 1, payload size might be so small that is not even worth it to waste CPU on compression.

Commit:
added error logging

Commit:
addressed CR comments, fixed UTs

vigneshashokan commented:
> Thanks for the PR.
> Can you please fix the tests?
> Can you please set appropriate header for gzip `req.Header.Set("Content-Encoding", "gzip")`
> I think it will be good if compression can be turned off optionally in configuration. In case someone wants real time logs and they limit batch size to 1, payload size might be so small that is not even worth it to waste CPU on compression.

I refactored code a bit, added error loggers and fixed the UTs. Turning off compression optionally is added work and I feel is out of scope for this issue. Work might be small, but we need to add more UTs for that as well, so I didn't do that. However I'm open to working on that in a different PR, if that's okay with you. Let me know what you think.

Commit:
added gzip in req header as content encoding

vigneshashokan commented:
Hey @mmarinm. Thanks for accepting my PR. If it isn't too much trouble, I would appreciate if you can add `hacktoberfest-accepted` label to this PR (https://hacktoberfest.digitalocean.com/hacktoberfest-update)
